### PR TITLE
Remove all PBX state from targets, improve project generation performance

### PR DESCRIPTION
**Pull request by dnkoutso**
`pod_target` and `target` generally hold on to too much state. These core models of the library try to represent a little bit of everything.

This PR removes all `PBX*` related properties from these classes and moves them all into a `TargetInstallationResult` container that basically contains all native targets that were created from an installation of a target.

Part of this PR also includes re-writing the `set_target_dependencies` method in `pods_project_generator` as it has become very hard to understand and also very inefficient as the same targets are being processed over and over again.

I have re-written all tests in `pods_project_generator` and have fixed various edge cases in the process.

Finally, within this PR, pod targets that have no sources (`should_build?` returns `false`) now instread get a placeholder `PBXAggregateTarget` that helps us wire dependencies better and avoids adding them into the aggregate targets set of dependencies.

Supersedes #https://github.com/CocoaPods/CocoaPods/pull/7603

/cc @jmkk

### CODE REVIEW

File: lib/cocoapods/installer.rb [hunk 1]
```diff
@@ -173,7 +173,7 @@ def download_dependencies
     private
 
     def create_generator
-      Xcode::PodsProjectGenerator.new(aggregate_targets, sandbox, pod_targets, analysis_result, installation_options, config)
+      Xcode::PodsProjectGenerator.new(sandbox, aggregate_targets, pod_targets, analysis_result, installation_options, config)
```
Comment (dnkoutso) [inline lib/cocoapods/installer.rb]: for consistency I moved `sandbox` to be first param, no changes.

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 2]
```diff
@@ -170,95 +177,122 @@ def install_libraries
...
         # links the pod targets among each other.
         #
+        # @param  [Array[Hash{String=>TargetInstallationResult}]] target_installation_results
+        #         the installation results array that was produced including pod target installation results and
+        #         aggregate target installation results.
+        #
         # @return [void]
         #
-        def set_target_dependencies
+        def wire_target_dependencies(target_installation_results)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: recommend side by side diff here, this method is verbose but it's in my opinion *way* simpler to understand, it also helps dramatically with performance for large projects.

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* [WIP] Remove all PBX state from targets, improve project generation performance  
  [dnkoutso](https://github.com/dnkoutso)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 3]
```diff
@@ -316,57 +341,44 @@ def add_test_targets
...
-                test_specification = file_accessor.spec.test_specification?
-
-                if test_specification
-                  target.test_resource_bundle_targets << bundle_target
-                else
-                  target.resource_bundle_targets << bundle_target
-                end
-
-                if target.should_build?
-                  native_target.add_dependency(bundle_target)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: this logic has moved into `pods_project_generator` for clarity. In the past we were creating resource bundle targets here during installation but we were also wiring the dependencies up. This made it hard to follow. This class is solely responsible for *creating* targets and configuring them, it should not do dependencies as it does not have enough information to do so.

This method has moved and a test has been added.

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 4]
```diff
@@ -62,9 +63,9 @@ def initialize(aggregate_targets, sandbox, pod_targets, analysis_result, install
         def generate!
           prepare
           install_file_references
-          install_libraries
-          integrate_targets
-          set_target_dependencies
+          @target_installation_results = install_libraries
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: does this need to be an ivar?

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 5]
```diff
@@ -75,12 +76,18 @@ def write
             if installation_options.deterministic_uuids?
               UI.message('- Generating deterministic UUIDs') { project.predictabilize_uuids }
             end
+            project.recreate_user_schemes(false)
             library_product_types = [:framework, :dynamic_library, :static_library]
+            pod_target_installation_results = @target_installation_results.first.values
             project.recreate_user_schemes(false) do |scheme, target|
+              next if target.is_a?(Xcodeproj::Project::Object::PBXAggregateTarget)
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: can this use `responds_to?`

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 6]
```diff
@@ -75,12 +76,18 @@ def write
             if installation_options.deterministic_uuids?
               UI.message('- Generating deterministic UUIDs') { project.predictabilize_uuids }
             end
+            project.recreate_user_schemes(false)
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: is this coming from somewhere else?

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 7]
```diff
@@ -170,95 +177,122 @@ def install_libraries
               pod_target.umbrella_header_path
             end.compact.group_by(&:dirname)
 
-            pod_targets.sort_by(&:name).each do |pod_target|
+            pod_target_installation_results = Hash[pod_targets.sort_by(&:name).map do |pod_target|
               target_installer = PodTargetInstaller.new(sandbox, pod_target, umbrella_headers_by_dir)
-              target_installer.install!
+              [pod_target.name, target_installer.install!]
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: any reason to key by the name instead of the target itself?

File: lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            test_native_targets.find do |native_target|
+              native_target.symbol_type == target.product_type_for_test_type(spec.test_type)
+            end
+          end
+
+          # @return [Hash{Array => Specification}] a hash where the keys are the test native targets and the value
+          #         an array of all the test specs associated with this native target.
+          #
+          def test_specs_by_native_target
+            target.test_specs.group_by do |test_spec|
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb]: this looks weird to me... maybe a loop + appending to a dictionary makes more sense here?

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 4 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: yes unfortunately for this PR. It seems that `#write` method depends on this result and someone to have called `generate!` first. 

I want to clean it up in a follow up PR if its OK.

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 6 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: whoops thats a mistake

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 7 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: uh no really I had initially fallen into the trap it wont be hashable...

File: lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb (hunk 8 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb]: it's copied from `pod_target` ¯\\\_(ツ)_/¯

File: lib/cocoapods/installer.rb [hunk 9]
```diff
@@ -200,6 +200,11 @@ def generate_pods_project(generator = create_generator)
     #
     attr_reader :analysis_result
 
+    # @return [Array<Hash{String, TargetInstallationResult}>] the installation results produced by the pods project
+    #         generator
+    #
+    attr_reader :target_installation_results
```
Comment (dnkoutso) [inline lib/cocoapods/installer.rb]: unfortunately to land this for validator changes I need the `installer` to hold on to this state. I do not want to fix this now since this is already blowing out of proportion in terms of changes. `installer` needs a clean up in a follow up.

**Comment by dnkoutso** [comment]
WIth this https://github.com/CocoaPods/cocoapods-integration-specs/pull/154 it goes green.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 10]
```diff
@@ -247,38 +269,40 @@ def add_test_app_host_targets
...
                 test_host << 'Contents/MacOS/' if target.platform == :osx
                 test_host << name.to_s
                 configuration.build_settings['TEST_HOST'] = test_host
               end
               target_attributes = project.root_object.attributes['TargetAttributes'] || {}
-              target_attributes[native_test_target.uuid.to_s] = { 'TestTargetID' => app_host_target.uuid.to_s }
+              target_attributes[test_native_target.uuid.to_s] = { 'TestTargetID' => app_host_target.uuid.to_s }
               project.root_object.attributes['TargetAttributes'] = target_attributes
-            end
+            end.compact
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: have to add `compact` because we now use `map` and line above includes `next unless...`

File: spec/unit/installer/xcode/pods_project_generator_spec.rb [hunk 11]
```diff
@@ -4,555 +4,280 @@ module Pod
...
+            user_target = stub('SampleApp-iOS-User-Target', :symbol_type => :tv_extension)
+            @ios_target.stubs(:user_targets).returns([user_target])
+            @generator.generate!
+            build_settings = @generator.project.targets.find { |t| t.name == 'Pods-SampleApp-iOS' }.build_configurations.map(&:build_settings)
+            build_settings.each do |build_setting|
+              build_setting['APPLICATION_EXTENSION_API_ONLY'].should == 'YES'
             end
+          end
 
-            describe 'sharing schemes of development pods' do
```
Comment (dnkoutso) [inline spec/unit/installer/xcode/pods_project_generator_spec.rb]: will add these tests again

**Comment by dnkoutso** [comment]
Greeeeen! but just need to add back some tests when I find time this week.

File: lib/cocoapods/installer.rb [hunk 12]
```diff
@@ -173,14 +173,14 @@ def download_dependencies
...
-      Xcode::PodsProjectGenerator.new(aggregate_targets, sandbox, pod_targets, analysis_result, installation_options, config)
+      Xcode::PodsProjectGenerator.new(sandbox, aggregate_targets, pod_targets, analysis_result, installation_options, config)
     end
 
     # Generate the 'Pods/Pods.xcodeproj' project.
     #
     def generate_pods_project(generator = create_generator)
       UI.section 'Generating Pods project' do
-        generator.generate!
+        @target_installation_results = generator.generate!
```
Comment (dnkoutso) [inline lib/cocoapods/installer.rb]: unfortunately and for this PR this will have to remain exposed.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 13]
```diff
@@ -25,33 +25,45 @@ def initialize(sandbox, project, target, umbrella_headers_by_dir = nil)
...
+                resource_bundle_targets = add_resources_bundle_targets(file_accessors)
+                return TargetInstallationResult.new(target, native_target, resource_bundle_targets)
+              end
+
               create_support_files_dir
+
+              native_target = add_target
+              resource_bundle_targets = add_resources_bundle_targets(file_accessors)
+
               if target.contains_test_specifications?
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: I wonder if this conditional is necessary, eg if `add_test_targets` handles not having any test targets correctly?

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 14]
```diff
@@ -25,33 +25,45 @@ def initialize(sandbox, project, target, umbrella_headers_by_dir = nil)
...
               end
-              add_resources_bundle_targets
-              add_files_to_build_phases
-              create_xcconfig_file
-              create_test_xcconfig_files if target.contains_test_specifications?
+
+              add_files_to_build_phases(native_target, test_native_targets)
+
+              create_xcconfig_file(native_target, resource_bundle_targets)
+              create_test_xcconfig_files(test_native_targets, test_resource_bundle_targets) if target.contains_test_specifications?
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: same as above

**Comment by dnkoutso** [comment]
-14 total lines of code diff.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 15]
```diff
@@ -25,33 +25,45 @@ def initialize(sandbox, project, target, umbrella_headers_by_dir = nil)
...
-            unless target.should_build?
-              add_resources_bundle_targets
-              return
-            end
-
             UI.message "- Installing target `#{target.name}` #{target.platform}" do
-              add_target
+              test_file_accessors, file_accessors = target.file_accessors.partition { |fa| fa.spec.test_specification? }
+
+              unless target.should_build?
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: this looks like it won't handle test specs if they exist?

File: lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+          end
+
+          # Returns the corresponding native target to use based on the provided specification.
+          #
+          # @param  [Specification] spec
+          #         The specification to base from in order to find the native target.
+          #
+          # @return [PBXNativeTarget] the native target to use or `nil` if none is found.
+          #
+          def native_target_for_spec(spec)
```
Comment (segiddins) [inline lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb]: I think this logic is duplicated above, we should use this method instead

File: spec/unit/validator_spec.rb [hunk 17]
```diff
@@ -1079,12 +1079,17 @@ def test_swiftpod_with_dot_swift_version(version = '3.1.0')
           debug_configuration_two = stub(:build_settings => {})
           native_target_one = stub(:build_configuration_list => stub(:build_configurations => [debug_configuration_one]))
           native_target_two = stub(:build_configuration_list => stub(:build_configurations => [debug_configuration_two]))
-          pod_target_one = stub(:uses_swift? => true, :swift_version => '4.0', :native_target => native_target_one, :test_native_targets => [])
-          pod_target_two = stub(:uses_swift? => true, :swift_version => '3.2', :native_target => native_target_two, :test_native_targets => [])
+          pod_target_one = stub(:name => 'PodTarget1', :uses_swift? => true, :swift_version => '4.0',
```
Comment (segiddins) [inline spec/unit/validator_spec.rb]: remove the old attributes from this

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb (hunk 13 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: hm it actually will! let me see

File: lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb (hunk 16 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb]: done

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 18]
```diff
@@ -170,109 +177,121 @@ def install_libraries
...
+              add_system_framework_dependencies(pod_target_installation_result_values)
             end
 
-            add_system_framework_dependencies
+            aggregate_target_installation_results = Hash[aggregate_targets.sort_by(&:name).map do |target|
+              target_installer = AggregateTargetInstaller.new(sandbox, @project, target)
+              [target.name, target_installer.install!]
+            end]
+
+            return [pod_target_installation_results, aggregate_target_installation_results]
```
Comment (DanToml) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: Could target installation results potentially be modelled as something clearer than a tuple array? - It's not super clear what is what without tracing code through to here what indexes are which.

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 19]
```diff
@@ -170,109 +177,121 @@ def install_libraries
...
-          pod_targets.select(&:should_build?).sort_by(&:name).each do |pod_target|
-            test_file_accessors, file_accessors = pod_target.file_accessors.partition { |fa| fa.spec.test_specification? }
-            file_accessors.each do |file_accessor|
-              add_system_frameworks_to_native_target(file_accessor, pod_target.native_target)
-            end
-            test_file_accessors.each do |test_file_accessor|
-              native_target = pod_target.native_target_for_spec(test_file_accessor.spec)
-              add_system_frameworks_to_native_target(test_file_accessor, native_target)
+        def add_system_framework_dependencies(pod_target_installation_results)
+          pod_target_installation_results.each do |target_installation_result|
```
Comment (DanToml) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: This was previously sorted by name, changing that will cause project diffs through having a different sort order for system frameworks. (Not necessarily a big deal, but it's a behaviour change)

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 2 above)
Comment (DanToml) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: This method is mostly unreadable now :/ - There are a few cases within it where things can be broken out into their own methods without meaningful perf impact, that would improve readability and make testing easier.

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 19 above)
Comment (DanToml) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: Also looks like this fn is missing test coverage

File: lib/cocoapods/user_interface.rb [hunk 20]
```diff
@@ -62,6 +62,7 @@ def section(title, verbose_prefix = '', relative_indentation = 0)
         self.indentation_level += relative_indentation
         self.title_level += 1
         yield if block_given?
+      ensure
```
Comment (DanToml) [inline lib/cocoapods/user_interface.rb]: +1 on this change, not clear why it's in this pr tho...

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 18 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: you are right, I didn't have a better way. Let me think about it and figure out if I can do so.

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 19 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: Will fix the sorting and yes test coverage was *always* missing from this method, I had removed the "TODO add specs" above. Will add tests.

File: lib/cocoapods/user_interface.rb (hunk 20 above)
Comment (dnkoutso) [reply lib/cocoapods/user_interface.rb]: +1 separate PR coming up.

Comment (dnkoutso) [reply lib/cocoapods/user_interface.rb]: PR https://github.com/CocoaPods/CocoaPods/pull/7634

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 19 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: test added. This was easier as the `OrangeFramework` pod I use depends on `s.frameworks 'UIKIt'` so I wrote one that ensures the native target includes this framework.

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 18 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: @DanToml best I came up with is just create a `Struct` that has the two values, its definitely more readable now and avoids using `first` and `last`.

Please let me know what you think

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 19 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: Fixed sorting
